### DISCUSSION TITLE
About porting SH8601 chip to Arduino_ GFX

### DISCUSSION THREAD

**Discussion by Xk-w** [moononournation/Arduino_GFX#344] [Q&A] [answered]
I want to port the SH8601 screen chip to Arduino_ There was a problem in the GFX library and it can be run using software, but using Arduino_ The ESP32QSPI method cannot be used to configure hardware, but the screen cannot be lit, but it can also send an invertDisplay (true); Function to make the screen brighter... Now I don't know, I don't know how to solve this problem next

Here is my code：

#include "Arduino_SH8601.h"

Arduino_SH8601::Arduino_SH8601(Arduino_DataBus *bus, int8_t rst, uint8_t r, bool ips, int16_t w, int16_t h)
    : Arduino_TFT(bus, rst, r, ips, w, h, 0, 0, 0, 0)
{
}

bool Arduino_SH8601::begin(int32_t speed)
{
    return Arduino_TFT::begin(speed);
}

void Arduino_SH8601::writeAddrWindow(int16_t x, int16_t y, uint16_t w, uint16_t h)
{
    if ((x != _currentX) || (w != _currentW))
    {
        _currentX = x;
        _currentW = w;
        x += _xStart;
        _bus->writeC8D16D16(SH8601_W_CASET, x, x + w - 1);
    }

    if ((y != _currentY) || (h != _currentH))
    {
        _currentY = y;
        _currentH = h;
        y += _yStart;
        _bus->writeC8D16D16(SH8601_W_PASET, y, y + h - 1);
    }

    // if ((x != _currentX) || (w != _currentW))
    // {
    //     int16_t x_start = x + _xStart, x_end = x + w - 1 + _xStart;

    //     _bus->writeCommand(SH8601_W_CASET); // Column addr set
    //     _bus->write(x_start >> 8);
    //     _bus->write(x_start & 0xFF); // XSTART
    //     _bus->write(x_end >> 8);
    //     _bus->write(x_end & 0xFF); // XEND

    //     _currentX = x;
    //     _currentW = w;
    // }
    // if ((y != _currentY) || (h != _currentH))
    // {
    //     int16_t y_start = y + _yStart, y_end = y + h - 1 + _yStart;

    //     _bus->writeCommand(SH8601_W_PASET); // Row addr set
    //     _bus->write(y_start >> 8);
    //     _bus->write(y_start & 0xFF); // YSTART
    //     _bus->write(y_end >> 8);
    //     _bus->write(y_end & 0xFF); // YEND

    //     _currentY = y;
    //     _currentH = h;
    // }

    // _bus->beginWrite();
    // _bus->writeCommand(SH8601_W_CASET);
    // _bus->write16(x);
    // _bus->write16(w);
    // _bus->endWrite();

    // _bus->beginWrite();
    // _bus->writeCommand(SH8601_W_PASET);
    // _bus->write16(y);
    // _bus->write16(h);
    // _bus->endWrite();

    _bus->writeCommand(SH8601_W_RAMWR); // write to RAM
}

/**************************************************************************/
/*!
    @brief   Set origin of (0,0) and orientation of TFT display
    @param   m  The index for rotation, from 0-3 inclusive
*/
/**************************************************************************/
void Arduino_SH8601::setRotation(uint8_t r)
{
    Arduino_TFT::setRotation(r);
    switch (_rotation)
    {
    case 0:
        // r = SH8601_MADCTL_RGB;
    case 1:
        r = SH8601_MADCTL_MV | SH8601_MADCTL_MX | SH8601_MADCTL_RGB;
        break;
    case 2:
        r = SH8601_MADCTL_MX | SH8601_MADCTL_MY | SH8601_MADCTL_RGB;
        break;
    case 3:
        r = SH8601_MADCTL_MV | SH8601_MADCTL_MX | SH8601_MADCTL_RGB;
        break;
    default:

        break;
    }
    _bus->beginWrite();
    _bus->writeC8D8(SH8601_W_MADCTL, r);
    _bus->endWrite();
}

void Arduino_SH8601::invertDisplay(bool i)
{
    _bus->sendCommand((_ips ^ i) ? SH8601_C_INVON : SH8601_C_INVOFF);
}

void Arduino_SH8601::displayOn(void)
{
    _bus->writeCommand(SH8601_C_SLPOUT);
    delay(SH8601_SLPOUT_DELAY);
}

void Arduino_SH8601::displayOff(void)
{
    _bus->sendCommand(SH8601_C_SLPIN);
    delay(SH8601_SLPIN_DELAY);
}

// Companion code to the above tables.  Reads and issues
// a series of LCD commands stored in PROGMEM byte array.
void Arduino_SH8601::tftInit()
{
    if (_rst != GFX_NOT_DEFINED)
    {
        pinMode(_rst, OUTPUT);
        digitalWrite(_rst, HIGH);
        delay(10);
        digitalWrite(_rst, LOW);
        delay(SH8601_RST_DELAY);
        digitalWrite(_rst, HIGH);
        delay(SH8601_RST_DELAY);
    }
    else
    {
        // Software Rest
        _bus->sendCommand(SH8601_C_SWRESET);
        delay(SH8601_RST_DELAY);
    }

    _bus->batchOperation(sh8601_init_operations, sizeof(sh8601_init_operations));

    // _bus->beginWrite();

    // _bus->writeCommand(SH8601_C_SLPOUT); // Sleep Out
    // delay(SH8601_SLPOUT_DELAY);

    // _bus->writeC8D16(SH8601_W_SETTSL, 0x01D1);

    // _bus->writeC8D8(SH8601_WC_TEARON, 0x00);
    // _bus->writeC8D8(SH8601_W_PIXFMT, 0x55);
    // _bus->writeC8D8(SH8601_W_WCTRLD1, 0x20);
    // delay(10);
    // _bus->writeCommand(SH8601_C_DISPON); // Display on
    // delay(10);
    // _bus->writeC8D8(SH8601_W_WDBRIGHTNESSVALNOR, 0xFF); // Write Display Brightness   MAX_VAL=0XFF
    // _bus->endWrite();

    // _bus->sendCommand(SH8601_C_SLPOUT); // Exit Sleep
    // delay(SH8601_SLPOUT_DELAY);

    // _bus->sendCommand(SH8601_W_SETTSL);
    // _bus->sendData(0x01);
    // _bus->sendData(0xD1);

    // _bus->sendCommand(SH8601_WC_TEARON);
    // _bus->sendData(0x00);

    // _bus->sendCommand(SH8601_W_PIXFMT);
    // _bus->sendData(0x55);

    // _bus->sendCommand(SH8601_W_WCTRLD1);
    // _bus->sendData(0x20);

    // _bus->sendCommand(SH8601_C_DISPON);

    // _bus->sendCommand(SH8601_W_WDBRIGHTNESSVALNOR);
    // _bus->sendData(0xFF);

    // writeAddrWindow(0,0,250,250);
    // for (int i = 0; i < 465; i++)
    // {
    //     for (int j = 0; j < 465; j++)
    //     {
    //         if (i == 465 / 2 - 1 || i == 465 / 2 || j == 465 / 2 - 1 || j == 465 / 2)
    //         {
    //             _bus->write(0xf8);
    //             _bus->write(0x00);
    //         }
    //         else
    //         {
    //             _bus->write(0xFF);
    //             _bus->write(0xFF);
    //         }
    //     }
    // }

    // writeAddrWindow(0, 0, 466, 466);
    // _bus->beginWrite();
    // for (int i = 0; i < 466; i++)
    // {
    //     for (int j = 0; j < 466; j++)
    //     {
    //         _bus->write(0x07);
    //         _bus->write(0xE0);
    //     }
    // }
    // _bus->endWrite();

    // invertDisplay(true);//Here you can run
    // delay(1000);
    // displayOff();
    // delay(1000);
    // displayOn();
    // delay(1000);
    // displayOff();
    // delay(1000);
    // displayOn();
}

Here is the initialization code：

#define SH8601_MAXWIDTH 480  ///< SH8601 max TFT width
#define SH8601_MAXHEIGHT 480 ///< SH8601 max TFT width

#define SH8601_RST_DELAY 200    ///< delay ms wait for reset finish
#define SH8601_SLPIN_DELAY 120  ///< delay ms wait for sleep in finish
#define SH8601_SLPOUT_DELAY 120 ///< delay ms wait for sleep out finish

// User Command
#define SH8601_C_NOP 0x00          // nop
#define SH8601_C_SWRESET 0x01      // Software Reset
#define SH8601_R_RDID 0x04         // Read Display Identification Information ID/1/2/3
#define SH8601_R_RDNERRORSDSI 0x05 // Read Number of Errors on DSI
#define SH8601_R_RDPOWERMODE 0x0A  // Read Display Power Mode
#define SH8601_R_RDMADCTL 0x0B     // Read Display MADCTL
#define SH8601_R_RDPIXFMT 0x0C     // Read Display Pixel Format
#define SH8601_R_RDIMGFMT 0x0D     // Read Display Image Mode
#define SH8601_R_RDSIGMODE 0x0E    // Read Display Signal Mode
#define SH8601_R_RDSELFDIAG 0x0F   // Read Display Self-Diagnostic Result

#define SH8601_C_SLPIN 0x10  // Sleep In
#define SH8601_C_SLPOUT 0x11 // Sleep Out
#define SH8601_C_PTLON 0x12  // Partial Display On
#define SH8601_C_NORON 0x13  // Normal Display mode on

#define SH8601_C_INVOFF 0x20  // Inversion Off
#define SH8601_C_INVON 0x21   // Inversion On
#define SH8601_C_ALLPOFF 0x22 // All pixels off
#define SH8601_C_ALLPON 0x23  // All pixels on
#define SH8601_C_DISPOFF 0x28 // Display off
#define SH8601_C_DISPON 0x29  // Display on
#define SH8601_W_CASET 0x2A   // Column Address Set
#define SH8601_W_PASET 0x2B   // Page Address Set
#define SH8601_W_RAMWR 0x2C   // Memory Write Start

#define SH8601_W_PTLAR 0x30   // Partial Area Row Set
#define SH8601_W_PTLAC 0x31   // Partial Area Column Set
#define SH8601_C_TEAROFF 0x34 // Tearing effect off
#define SH8601_WC_TEARON 0x35 // Tearing effect on
#define SH8601_W_MADCTL 0x36  // Memory data access control
#define SH8601_C_IDLEOFF 0x38 // Idle Mode Off
#define SH8601_C_IDLEON 0x39  // Idle Mode On
#define SH8601_W_PIXFMT 0x3A  // Write Display Pixel Format
#define SH8601_W_WRMC 0x3C    // Memory Write Continue

#define SH8601_W_SETTSL 0x44             // Write Tearing Effect Scan Line
#define SH8601_R_GETSL 0x45              // Read Scan Line Number
#define SH8601_C_SPIROFF 0x46            // SPI read Off
#define SH8601_C_SPIRON 0x47             // SPI read On
#define SH8601_C_AODMOFF 0x48            // AOD Mode Off
#define SH8601_C_AODMON 0x49             // AOD Mode On
#define SH8601_W_WDBRIGHTNESSVALAOD 0x4A // Write Display Brightness Value in AOD Mode
#define SH8601_R_RDBRIGHTNESSVALAOD 0x4B // Read Display Brightness Value in AOD Mode
#define SH8601_W_DEEPSTMODE 0x4F         // Deep Standby Mode On

#define SH8601_W_WDBRIGHTNESSVALNOR 0x51 // Write Display Brightness Value in Normal Mode
#define SH8601_R_RDBRIGHTNESSVALNOR 0x52 // Read display brightness value in Normal Mode
#define SH8601_W_WCTRLD1 0x53            // Write CTRL Display1
#define SH8601_R_RCTRLD1 0x54            // Read CTRL Display1
#define SH8601_W_WCTRLD2 0x55            // Write CTRL Display2
#define SH8601_R_RCTRLD2 0x56            // Read CTRL Display2
#define SH8601_W_WCE 0x58                // Write CE
#define SH8601_R_RCE 0x59                // Read CE

#define SH8601_W_WDBRIGHTNESSVALHBM 0x63 // Write Display Brightness Value in HBM Mode
#define SH8601_R_WDBRIGHTNESSVALHBM 0x64 // Read Display Brightness Value in HBM Mode
#define SH8601_W_WHBMCTL 0x66            // Write HBM Control

#define SH8601_W_COLORSET0 0x70  // Color Set 0
#define SH8601_W_COLORSET1 0x71  // Color Set 1
#define SH8601_W_COLORSET2 0x72  // Color Set 2
#define SH8601_W_COLORSET3 0x73  // Color Set 3
#define SH8601_W_COLORSET4 0x74  // Color Set 4
#define SH8601_W_COLORSET5 0x75  // Color Set 5
#define SH8601_W_COLORSET6 0x76  // Color Set 6
#define SH8601_W_COLORSET7 0x77  // Color Set 7
#define SH8601_W_COLORSET8 0x78  // Color Set 8
#define SH8601_W_COLORSET9 0x79  // Color Set 9
#define SH8601_W_COLORSET10 0x7A // Color Set 10
#define SH8601_W_COLORSET11 0x7B // Color Set 11
#define SH8601_W_COLORSET12 0x7C // Color Set 12
#define SH8601_W_COLORSET13 0x7D // Color Set 13
#define SH8601_W_COLORSET14 0x7E // Color Set 14
#define SH8601_W_COLORSET14 0x7F // Color Set 15

#define SH8601_W_COLOROPTION 0x80 // Color Option

#define SH8601_R_RDDBSTART 0xA1         // Read DDB start
#define SH8601_R_DDBCONTINUE 0xA8       // Read DDB Continue
#define SH8601_R_RFIRCHECKSUN 0xAA      // Read First Checksum
#define SH8601_R_RCONTINUECHECKSUN 0xAF // Read Continue Checksum

#define SH8601_W_SPIMODECTL 0xC4 // SPI mode control

#define SH8601_R_RDID1 0xDA // Read ID1
#define SH8601_R_RDID2 0xDB // Read ID2
#define SH8601_R_RDID3 0xDC // Read ID3

// parameters
#define SH8601_MADCTL_MY 0x80  // Bottom to top
#define SH8601_MADCTL_MX 0x40  // Right to left
#define SH8601_MADCTL_MV 0x20  // Reverse Mode
#define SH8601_MADCTL_ML 0x10  // LCD refresh Bottom to top
#define SH8601_MADCTL_RGB 0x00 // Red-Green-Blue pixel order
#define SH8601_MADCTL_BGR 0x08 // Blue-Green-Red pixel order
#define SH8601_MADCTL_MH 0x04  // LCD refresh right to left
#define SH8601_MADCTL_SS 0x02
#define SH8601_MADCTL_GS 0x01

static const uint8_t sh8601_init_operations[] = {

    BEGIN_WRITE,
    WRITE_COMMAND_8, SH8601_C_SLPOUT,
    END_WRITE,

    DELAY, SH8601_SLPOUT_DELAY,

    BEGIN_WRITE,
    // WRITE_COMMAND_8, SH8601_W_CASET,
    // WRITE_BYTES, 4,
    // 0x00,
    // 0x00, // 0  Start of Column
    // 0x01,
    // 0xD1, // 465+1  End of Column

    // WRITE_COMMAND_8, SH8601_W_PASET,
    // WRITE_BYTES, 4,
    // 0x00,
    // 0x00, // 0  Start of Column
    // 0x01,
    // 0xD1, // 465+1  End of Column

    WRITE_C8_D16, SH8601_W_SETTSL, 0x01, 0xD1,

    WRITE_C8_D8, SH8601_WC_TEARON, 0x00,

    WRITE_C8_D8, SH8601_W_PIXFMT, 0x55,

    WRITE_C8_D8, SH8601_W_WCTRLD1, 0x20, // Brightness Control On
    // END_WRITE,

    // DELAY, 10,

    // BEGIN_WRITE,
    // WRITE_C8_D8, SH8601_W_SPIMODECTL, 0x80,

    WRITE_C8_D8, SH8601_W_WDBRIGHTNESSVALNOR, 0x00,
    // END_WRITE,

    // DELAY, 10,

    // BEGIN_WRITE,
    WRITE_COMMAND_8, SH8601_C_DISPON,
    // END_WRITE,

    // DELAY, 10,

    WRITE_C8_D8, SH8601_W_WDBRIGHTNESSVALNOR, 0xFF,

    END_WRITE,

    DELAY, 10};

The following is the initialization code for the software driver：

void LCD_Init(void)
{
    digitalWrite(LCD_RST, HIGH);
    delay(10);
    digitalWrite(LCD_RST, LOW);
    delay(200);
    digitalWrite(LCD_RST, HIGH);
    delay(200);

    // write_cmd(0xFE);
    // write_dat(0x01);
    // digitalWrite(LCD_CS, HIGH);
    // delayMicroseconds(4);

    // inital code version date: 211026
    // Truly1.78"+SH8601A RM20013

    digitalWrite(LCD_CS, HIGH);
    digitalWrite(LCD_CS, LOW);
    WC(0x11);
    delay(120);
    // digitalWrite(LCD_CS, HIGH);
    // digitalWrite(LCD_CS, LOW);
    // WC(0x2A);
    // WD(0x00);
    // WD(0x00);
    // WD(0x01);
    // WD(0xC5);
    // digitalWrite(LCD_CS, HIGH);
    // digitalWrite(LCD_CS, LOW);
    // WC(0x2B);
    // WD(0x00);
    // WD(0x00);
    // WD(0x01);
    // WD(0xC5);
    digitalWrite(LCD_CS, HIGH);
    digitalWrite(LCD_CS, LOW);
    WC(0x44);
    WD(0x01);
    WD(0xD1);
    digitalWrite(LCD_CS, HIGH);
    digitalWrite(LCD_CS, LOW);
    WC(0x35);
    WD(0x00);
    digitalWrite(LCD_CS, HIGH);
    digitalWrite(LCD_CS, LOW);
    WC(0x3a);
    WD(0x55);
    digitalWrite(LCD_CS, HIGH);
    digitalWrite(LCD_CS, LOW);
    WC(0x53);
    WD(0x20);
    digitalWrite(LCD_CS, HIGH);
    // delay(10);
    digitalWrite(LCD_CS, LOW);
    // WC(0xC4);
    // WD(0x80);
    WC(0x51);
    WD(0x00);
    digitalWrite(LCD_CS, HIGH);
    // delay(10);
    digitalWrite(LCD_CS, LOW);
    WC(0x29);
    digitalWrite(LCD_CS, HIGH);
    // delay(10);
    digitalWrite(LCD_CS, LOW);
    WC(0x51);
    WD(0xFF);
    digitalWrite(LCD_CS, HIGH);
}

**Comment by Xk-w** [comment, marked as answer]
The problem has been solved, it's a memory allocation issue

**Reply by beyond-space21** [reply]
hi i have a SH8601 based QSPI display.

I have problem in driving the display. please help me with a library or source code to drive the display

**Comment by Xk-w** [comment]
> hi i have a SH8601 based QSPI display.
> 
> I have problem in driving the display. please help me with a library or source code to drive the display

The following link can help you: [here](https://github.com/Xinyuan-LilyGO/T-Encoder-Pro)

**Reply by beyond-space21** [reply]
This repository would help me a lot. thankyou,

i have an initialization code given by the manufracturer

> R2A 00 00 01 6F 
R2B 00 00 01 BF 
R44 01 BF 
R36 02 
R35 00 
R3A 55 
R53 20 
//DispColor(WHITE)	
//RC0 5A 5A 
//RBA 0x81 
//UCS
R11,0x00 
Delayms(5)
Delayms(50)
R29  00

How to use this in arduino_sh8601.h?

Do you have any advise?

thnakyou for your support.

**Comment by Xk-w** [comment]
> This repository would help me a lot. thankyou,
> 
> i have an initialization code given by the manufracturer
> 
> > R2A 00 00 01 6F
> > R2B 00 00 01 BF
> > R44 01 BF
> > R36 02
> > R35 00
> > R3A 55
> > R53 20
> > //DispColor(WHITE)
> > //RC0 5A 5A
> > //RBA 0x81
> > //UCS
> > R11,0x00
> > Delayms(5)
> > Delayms(50)
> > R29  00
> 
> How to use this in arduino_sh8601.h?
> 
> Do you have any advise?
> 
> thnakyou for your support.

If you wish to modify the initialization code for SH8601, please make the changes according to the format described in the Arduino_GFX library. Specifically, you only need to modify the static const uint8_t sh8601_init_operations[] array.


The commands to be written can be referred to in the "Arduino_DataBus.cpp" file, specifically the void Arduino_DataBus::batchOperation(const uint8_t *operations, size_t len) function.


Based on the code provided by the manufacturer that you've given me, I can only provide a preliminary example:


R2A 00 00 01 6F 
I speculate that it should be written as follows: 
-> 
static const uint8_t sh8601_init_operations[] = {
BEGIN_WRITE, 
WRITE_COMMAND_8, 0x2A, 
WRITE_BYTES, 4, 0x00, 0x00, 0x01, 0x6F, 
...... (Omitted) 
END_WRITE,
};


The omitted code needs to be supplemented by yourself.
@beyond-space21
